src/solvers/flattening/boolbv_unary_minus.cpp

```diff
@@ -23,66 +25,53 @@ bvt boolbvt::convert_unary_minus(const unary_exprt &expr)
   if(width==0)
     return conversion_failed(expr);
 
-  const exprt::operandst &operands=expr.operands();
-
-  if(operands.size()!=1)
-    throw "unary minus takes one operand";
-
-  const exprt &op0=expr.op0();
+  const exprt &op = expr.op();
 
-  const bvt &op_bv=convert_bv(op0);
+  const bvt &op_bv = convert_bv(op, width);
 
   bvtypet bvtype=get_bvtype(type);
-  bvtypet op_bvtype=get_bvtype(op0.type());
-  std::size_t op_width=op_bv.size();
+  bvtypet op_bvtype = get_bvtype(op.type());
 
   bool no_overflow=(expr.id()=="no-overflow-unary-minus");
 
-  if(op_width==0 || op_width!=width)
-    return conversion_failed(expr);
-
   if(bvtype==bvtypet::IS_UNKNOWN &&
      (type.id()==ID_vector || type.id()==ID_complex))
   {
     const typet &subtype=ns.follow(type.subtype());
 
     std::size_t sub_width=boolbv_width(subtype);
 
-    if(sub_width==0 || width%sub_width!=0)
-      throw "unary-: unexpected vector operand width";
+    INVARIANT(
+      sub_width > 0,
+      "bitvector representation of type needs to have at least one bit");
+
+    INVARIANT(
+      width % sub_width == 0,
+      "total bitvector width needs to be a multiple of the component bitvector "
+      "widths");
 
-    std::size_t size=width/sub_width;
     bvt bv;
-    bv.resize(width);
 
-    for(std::size_t i=0; i<size; i++)
+    for(std::size_t sub_idx = 0; sub_idx < width; sub_idx += sub_width)
     {
       bvt tmp_op;
-      tmp_op.resize(sub_width);
 
-      for(std::size_t j=0; j<tmp_op.size(); j++)
-      {
-        assert(i*sub_width+j<op_bv.size());
-        tmp_op[j]=op_bv[i*sub_width+j];
-      }
-
-      bvt tmp_result;
+      const auto sub_it = std::next(op_bv.begin(), sub_idx);
+      std::copy_n(sub_it, sub_width, std::back_inserter(tmp_op));
```

Comment:
I'm sure this is a good idea, but seems unrelated to the goal of the commit?

Reply:
Ok, this has become a separate commit now.